tests/Neo.Plugins.Storage.Tests/StoreTest.cs

```diff
@@ -70,40 +94,98 @@ private void TestStorage(IStore store)
+            // Read
+
+            using (var store = plugin.GetStore())
+            {
+                var ret = store.TryGet(byte.MaxValue, new byte[] { 0x01, 0x02, 0x03 });
+                CollectionAssert.AreEqual(new byte[] { 0x04, 0x05, 0x06 }, ret);
+            }
+
+            // Delete
+
+            using (var store = plugin.GetStore())
             {
                 store.Delete(byte.MaxValue, new byte[] { 0x01, 0x02, 0x03 });
             }
+
+            // Read
+
+            using (var store = plugin.GetStore())
+            {
+                var ret = store.TryGet(byte.MaxValue, new byte[] { 0x01, 0x02, 0x03 });
+                Assert.IsNull(ret);
+            }
         }
 
         /// <summary>
-        /// Test Read
+        /// Test snapshot
         /// </summary>
-        /// <param name="store">Store</param>
-        /// <param name="shouldExist">Should exist</param>
-        private void TestPersistenceRead(IStore store, bool shouldExist)
+        /// <param name="plugin">Plugin</param>
+        /// <param name="commit">Commit</param>
+        private void TestPersistenceSnapshot(IStoragePlugin plugin, bool commit)
         {
-            using (store)
+            var key = new byte[] { 0x01, 0x02, 0x03 };
+            var value = new byte[] { 0x04, 0x05, 0x06 };
+
+            using (var store = plugin.GetStore())
             {
-                var ret = store.TryGet(byte.MaxValue, new byte[] { 0x01, 0x02, 0x03 });
+                store.Put(0x10, key, value);
+
+                using (var snapshot = store.GetSnapshot())
+                {
+                    snapshot.Delete(0x10, key);
+                    if (commit) snapshot.Commit();
+                }
+
+                if (commit)
+                {
+                    Assert.IsNull(store.TryGet(0x10, key));
+                }
+                else
+                {
+                    CollectionAssert.AreEqual(value, store.TryGet(0x10, key));
+                }
+            }
+
+            // Repeat with other store
+
+            using (var store = plugin.GetStore())
+            {
+                if (commit)
+                {
+                    Assert.IsNull(store.TryGet(0x10, key));
+                }
+                else
+                {
+                    CollectionAssert.AreEqual(value, store.TryGet(0x10, key));
+                }
+            }
+
+            // Try to write during a snapshot
+
+            using (var store = plugin.GetStore())
+            {
+                store.Put(0x10, key, value);
+
+                using (var snapshot = store.GetSnapshot())
+                {
+                    // Create a dummy snapshot and remove from the storage
+
+                    store.Delete(0x10, key);
+                }
 
-                if (shouldExist) CollectionAssert.AreEqual(new byte[] { 0x04, 0x05, 0x06 }, ret);
-                else Assert.IsNull(ret);
+                Assert.IsNull(store.TryGet(0x10, key));
```

Comment:
I think the non-null result occurs because you recover the db when disposing the snapshot.